fileutils-3.16/man/mv.1
.TH MV 1 "GNU File Utilities" "FSF" \" -*- nroff -*-
.SH NAME
mv \- rename files
.SH SYNOPSIS
.B mv
[options] source dest
.br
.B mv
[options] source... directory
.br
Options:
.br
[\-bfiuv] [\-S backup-suffix] [\-V {numbered,existing,simple}]
[\-\-backup] [\-\-force] [\-\-interactive] [\-\-update] [\-\-verbose]
[\-\-suffix=backup-suffix] [\-\-version-control={numbered,existing,simple}]
[\-\-help] [\-\-version]
.SH DESCRIPTION
This documentation is no longer being maintained and may be inaccurate
or incomplete.  The Texinfo documentation is now the authoritative source.
.PP
This manual page
documents the GNU version of
.BR mv .
If the last argument names an existing directory,
.B mv
moves each other
given file into a file with the same name in that directory.
Otherwise, if only two files are given, it moves the first onto the
second.  It is an error if the last argument is not a directory and
more than two files are given.  It can move only regular files across
filesystems.
.P
If a destination file is unwritable, the standard input is a tty, and
the \fI\-f\fR or \fI\-\-force\fR option is not given,
.B mv
prompts the user for whether to overwrite the file.  If the response
does not begin with `y' or `Y', the file is skipped.
.SS OPTIONS
.TP
.I "\-b, \-\-backup"
Make backups of files that are about to be removed.
.TP
.I "\-f, \-\-force"
Remove existing destination files and never prompt the user.
.TP
.I "\-i, \-\-interactive"
Prompt whether to overwrite each destination file that already
exists.  If the response does not begin with `y' or `Y', the file is
skipped.
.TP
.I "\-u, \-\-update"
Do not move a nondirectory that has an existing destination with the
same or newer modification time.
.TP
.I "\-v, \-\-verbose"
Print the name of each file before moving it.
.TP
.I "\-\-help"
Print a usage message on standard output and exit successfully.
.TP
.I "\-\-version"
Print version information on standard output then exit successfully.
.TP
.I "\-S, \-\-suffix backup-suffix"
The suffix used for making simple backup files can be set with the
.B SIMPLE_BACKUP_SUFFIX
environment variable, which can be overridden by this option.  If
neither of those is given, the default is `~', as it is in Emacs.
.TP
.I "\-V, \-\-version-control {numbered,existing,simple}"
The type of backups made can be set with the
.B VERSION_CONTROL
environment variable, which can be overridden by this option.  If
.B VERSION_CONTROL
is not set and this option is not given, the default backup type is
`existing'.  The value of the
.B VERSION_CONTROL
environment variable and the argument to this option are like the GNU
Emacs `version-control' variable; they also recognize synonyms that
are more descriptive.  The valid values are (unique abbreviations are
accepted):
.RS
.TP
`t' or `numbered'
Always make numbered backups.
.TP
`nil' or `existing'
Make numbered backups of files that already
have them, simple backups of the others.
.TP
`never' or `simple'
Always make simple backups.


zsh/ChangeLog
beta 0.75
Wed Jun 14 14:47:46 2000    <amold@SKYNYRD>
	* ntport\dirent.c: Fix serious bug in wildcard expansion on Win9x 
beta 0.74
Mon Jan 31 14:53:06 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed bug in converting '\'s in $HOME 
	to '/'s.
beta 0.73
Mon Jan 17 11:55:15 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed crash in long arglist introduced by
 	0.70
beta 0.72
Tue Jan 11 08:28:29 2000    <amold@SKYNYRD>
	* ntport\support.c: Fixed memory corruption on Win9x
beta 0.71
Mon Jan 10 11:49:39 2000    <amold@SKYNYRD>
	* ntport\stdio.c: Fix bug in inheritance of file handles.  
beta 0.69,0.70
Fri Dec 17 11:53:19 1999    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in argv[0] of spawned argument. Might
	have manifested as applications in ZSHROOT failing to spawn 
	themselves. 
	Also added winntnoquoteprotect option. Use with tools like MKS
	if you see unexpected quoting of "
beta 0.68
Mon Sep 27 11:20:24 1999    <amold@SKYNYRD>
	* ntport\support.c: support paths > 1024 characters. 
beta 0.67
Tue Sep 07 10:05:02 1999    <amold@SKYNYRD>
	* ntport\io.c: Fix bug that affected FPATH/autoload 
beta 0.65
Fri Aug 27 11:41:58 1999    <amold@SKYNYRD>
	* ntport\dirent.c: Fix completion with drive letters in string. 
beta 0.64
Fri Aug 27 11:04:11 1999    <amold@SKYNYRD>
	* ntport\io.c: Unfix Win95 spurious linefeed fix for Windows 98. 
beta 0.63
Wed May 19 16:21:37 1999    <amold@SKYNYRD>
	* exec.c, misc: VC 6.0 compliation errors. Disable optmizations
	in functions that call fork() to avoid losing global variables
	that might be in registers.
beta 0.62
Tue Apr 13 17:06:12 1999    <amold@SKYNYRD>
	* ntport\support.c, zle_misc.c: Keep server name in prompt
 	if %C is specified and pwd is UNC.
beta 0.61
Tue Mar 23 12:17:50 1999  unknown  <bogus@SKYNYRD>
	* ntport\support.c: bug fix for previous fix. 
beta 0.60
Sat Feb 13 17:49:14 1999    <amold@SKYNYRD>
	* ntport\support.c,ntbin.c: Fix lame arg-concatenation in
 	exec and start paths. commands with large argument strings should
	now execute *much* faster.
beta 0.57
Tue Feb 02 18:09:39 1999    <amold@SKYNYRD>
	* ntport\ntbin.c: fixed bug that improperly passed arguments
	to applications launched by windows "shell" associations. 
	For example, .bat files.
beta 0.55
Mon Oct 26 11:09:44 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: start builtin will not use '/' for option
 	flags. only '-' will work.
beta 0.54
Fri Oct 16 12:52:08 1998    <amold@SKYNYRD>
	* ntport\support.c: added WINNTWAITFORGUIAPPS option. 
beta 0.53
Mon Oct 12 12:31:59 1998    <amold@SKYNYRD>
	* init.c: look for "zshrc" as well as ".zshrc" 
beta 0.52
Mon Sep 28 17:31:08 1998    <amold@SKYNYRD>
	* utils.c: attempt to ignore case while correcting spelling
 	if WINNTIGNORECASE is set.
beta 0.51
Thu Sep 17 13:48:27 1998    <amold@SKYNYRD>
	* hashtable.c,support.c: filenames without extensions and
 	*only* files with extensions specified in the PATHEXT variable
 	will be hashed. The default extensions are ".com;.exe;.cmd;.bat"
beta 0.50
Wed Sep 16 15:58:56 1998    <amold@SKYNYRD>
	* hashtable.c: added hashing for external exes without
	extension. bogus spelling corrections should be reduced.
beta 0.49
Tue Sep 15 11:09:31 1998    <amold@SKYNYRD>
	* support.c: The shell will automatically set TERM to vt100
 	if it's not set in the environment. There should be no need
	to set the variable explicitly any more.
beta 0.48
Mon Aug 03 17:47:08 1998    <amold@SKYNYRD>
	* support.c: Fixed bug in converting paths to '/'-delimited 
beta 0.47
Mon Aug 03 16:59:38 1998    <amold@SKYNYRD>
	* utils.c: Fixed bug in applying previous patch. If %~ wasn't
	working, this should fix it.
beta 0.46
Sat Jul 18 14:36:10 1998    <amold@SKYNYRD>
	* utils.c,globals.h,zsh.h: patch from dominik for 
	winntconvertbackslash.
beta 0.45
Mon Jun 29 11:30:55 1998    <amold@SKYNYRD>
	* utils.c: patch from dominik vogt for honoring winntignorecase
 	in %~ prompt.
beta 0.44
Mon Jun 22 11:19:12 1998    <amold@SKYNYRD>
	* support.c,io.c: moved some startup code around. fixed spurious
 	linefeeds on win95
beta 0.43
Tue Jun 09 17:25:29 1998    <amold@SKYNYRD>
	* builtin.c,ntport\io.c: Fix for the read builtin not
 	echoing characters
beta 0.429
Thu Jun 04 17:43:54 1998    <amold@SKYNYRD>
	* utils.c: Fixed %~ prompt spec.
beta 0.428
Thu Jun 04 16:50:21 1998    <amold@SKYNYRD>
	* builtin.c: Fixed problems with the setting of PWD.
beta 0.427
Thu May 28 14:31:28 1998    <amold@SKYNYRD>
	* ntport\support.c: Fix to let apps with spaces in their path
	run. c:\program files\myapp.exe , for example.
beta 0.426
Wed May 27 10:46:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed stupid command buffer overflow
 	problem. If you crashed while grepping through a bunch of
	files, this should fix it.
beta 0.425
Fri May 22 08:55:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in parsing args to child processes.
beta 0.424
Sun May 17 13:18:21 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug that caused zsh to spawn a new
	console if $path contained a component '/'. 
beta 0.423
Fri May 15 16:08:17 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: winntlamepathfix now works with the 'start'
	builtin.
beta 0.421
Mon Apr 06 16:18:20 1998    <amold@SKYNYRD>
	* ntport\support.c,fork.c: Fixed bug in heap initialization for
	forked children that crashed the child process.
beta 0.42
Sun Apr 05 13:36:49 1998    <amold@SKYNYRD>
	* ntport\support.c: Put in code to fix the case of a directory
 	you cd into. This is extremely expensive and really silly, so
	it may become optional in the future.
beta 0.41
Sat Apr 04 15:27:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug that restricted command line
 	length
beta 0.401
Thu Mar 26 08:21:21 1998    <amold@SKYNYRD>
	* ntport\support.c: Fix to stop home from being re-set in child
 	if unset in parent. Also HOME will not be set by zsh if ZDOTDIR

fileutils-3.16/src/mv.c
/* mv -- move or rename files
   Copyright (C) 86, 89, 90, 91, 95, 1996 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* Options:
   -f, --force		Assume a 'y' answer to all questions it would
			normally ask, and not ask the questions.

   -i, --interactive	Require confirmation from the user before
			performing any move that would destroy an
			existing file.

   -u, --update		Do not move a nondirectory that has an
			existing destination with the same or newer
			modification time.

   -v, --verbose		List the name of each file as it is moved, and
			the name it is moved to.

   -b, --backup
   -S, --suffix
   -V, --version-control
			Backup file creation.

   Written by Mike Parker and David MacKenzie */

#ifdef _AIX
 #pragma alloca
#endif


#include <config.h>
#include <stdio.h>
#include <getopt.h>
#include <sys/types.h>

#include "system.h"
#include "path-concat.h"
#include "backupfile.h"
#include "error.h"

#ifdef HAVE_LCHOWN
# define chown(PATH, OWNER, GROUP) lchown(PATH, OWNER, GROUP)
#endif

#ifndef _POSIX_VERSION
uid_t geteuid ();
#endif

char *basename ();
enum backup_type get_version ();
int isdir ();
int yesno ();
int safe_read ();
int full_write ();
void strip_trailing_slashes ();
int euidaccess ();

// kms
#define rename rename

/* The name this program was run with. */
char *program_name;

/* If nonzero, query the user before overwriting files. */
static int interactive;

/* If nonzero, do not query the user before overwriting unwritable
   files. */
static int override_mode;

/* If nonzero, do not move a nondirectory that has an existing destination
   with the same or newer modification time. */
static int update = 0;

/* If nonzero, list each file as it is moved. */
static int verbose;

/* If nonzero, stdin is a tty. */
static int stdin_tty;

/* This process's effective user ID.  */
static uid_t myeuid;

/* FIXME */
static struct stat dest_stats, source_stats;

/* If nonzero, display usage information and exit.  */
static int show_help;

/* If nonzero, print the version on standard output and exit.  */
static int show_version;

static struct option const long_options[] =
{
  {"backup", no_argument, NULL, 'b'},
  {"force", no_argument, NULL, 'f'},
  {"interactive", no_argument, NULL, 'i'},
  {"suffix", required_argument, NULL, 'S'},
  {"update", no_argument, &update, 1},
  {"verbose", no_argument, &verbose, 1},
  {"version-control", required_argument, NULL, 'V'},
  {"help", no_argument, &show_help, 1},
  {"version", no_argument, &show_version, 1},
  {NULL, 0, NULL, 0}
};

/* If PATH is an existing directory, return nonzero, else 0.  */

static int
is_real_dir (const char *path)
{
  struct stat stats;

  return lstat (path, &stats) == 0 && S_ISDIR (stats.st_mode);
}

/* Copy regular file SOURCE onto file DEST.
   Return 1 if an error occurred, 0 if successful. */

static int
copy_reg (const char *source, const char *dest)
{
  int ifd;
  int ofd;
  char buf[1024 * 8];
  int len;			/* Number of bytes read into `buf'. */

  lstat (source, &source_stats);
  if (!S_ISREG (source_stats.st_mode))
    {
      error (0, 0,
	     _("cannot move `%s' across filesystems: Not a regular file"),
	     source);
      return 1;
    }

  if (unlink (dest) && errno != ENOENT)
    {
      error (0, errno, _("cannot remove `%s'"), dest);
      return 1;
    }

  ifd = open (source, O_RDONLY|O_BINARY, 0);
  if (ifd < 0)
    {
      error (0, errno, "%s", source);
      return 1;
    }
  ofd = open (dest, O_WRONLY | O_CREAT | O_TRUNC|O_BINARY, 0600);
  if (ofd < 0)
    {
      error (0, errno, "%s", dest);
      close (ifd);
      return 1;
    }

  while ((len = safe_read (ifd, buf, sizeof (buf))) > 0)
    {
      if (full_write (ofd, buf, len) < 0)
	{
	  error (0, errno, "%s", dest);
	  close (ifd);
	  close (ofd);
	  unlink (dest);
	  return 1;
	}
    }
  if (len < 0)
    {
      error (0, errno, "%s", source);
      close (ifd);
      close (ofd);
      unlink (dest);
      return 1;
    }

  if (close (ifd) < 0)
    {
      error (0, errno, "%s", source);
      close (ofd);
      return 1;
    }
  if (close (ofd) < 0)
    {
      error (0, errno, "%s", dest);
      return 1;
    }

  /* chown turns off set[ug]id bits for non-root,
     so do the chmod last.  */

  /* Try to copy the old file's modtime and access time.  */
  {
    struct utimbuf tv;

    tv.actime = source_stats.st_atime;
    tv.modtime = source_stats.st_mtime;
    if (utime (dest, &tv))
      {
	error (0, errno, "%s", dest);
	return 1;
      }
  }

  /* Try to preserve ownership.  For non-root it might fail, but that's ok.
     But root probably wants to know, e.g. if NFS disallows it.  */
  if (chown (dest, source_stats.st_uid, source_stats.st_gid)
      && (errno != EPERM || myeuid == 0))
    {
      error (0, errno, "%s", dest);
      return 1;
    }

  if (chmod (dest, source_stats.st_mode & 07777))
    {
      error (0, errno, "%s", dest);
      return 1;
    }

  return 0;
}

/* Move SOURCE onto DEST.  Handles cross-filesystem moves.
   If DEST is a directory, SOURCE must be also.
   Return 0 if successful, 1 if an error occurred.  */

static int
do_move (const char *source, const char *dest)
{
  char *dest_backup = NULL; int i;

  if (lstat (source, &source_stats) != 0)
    {
      error (0, errno, "%s", source);
      return 1;
    }

  if (lstat (dest, &dest_stats) == 0)
    {
      if (source_stats.st_dev == dest_stats.st_dev
	  && source_stats.st_ino == dest_stats.st_ino)
	{
	  error (0, 0, _("`%s' and `%s' are the same file"), source, dest);
	  return 1;
	}

      if (S_ISDIR (dest_stats.st_mode))
	{
	  error (0, 0, _("%s: cannot overwrite directory"), dest);
	  return 1;
	}

      if (!S_ISDIR (source_stats.st_mode) && update
	  && source_stats.st_mtime <= dest_stats.st_mtime)
	return 0;

      if (!override_mode && (interactive || stdin_tty))
// && euidaccess (dest, W_OK))
	{
	  fprintf (stderr, _("%s: replace `%s', overriding mode %04o? "),
		   program_name, dest,
		   (unsigned int) (dest_stats.st_mode & 07777));
	  if (!yesno ())
	    return 0;
	}
      else if (interactive)
	{
	  fprintf (stderr, _("%s: replace `%s'? "), program_name, dest);
	  if (!yesno ())
	    return 0;
	}

      if (backup_type != none)
	{
	  char *tmp_backup = find_backup_file_name (dest);
	  if (tmp_backup == NULL)
	    error (1, 0, _("virtual memory exhausted"));
	  dest_backup = (char *) alloca (strlen (tmp_backup) + 1);
	  strcpy (dest_backup, tmp_backup);
	  free (tmp_backup);
	  if (rename (dest, dest_backup))
	    {
	      if (errno != ENOENT)
		{
		  error (0, errno, _("cannot backup `%s'"), dest);
		  return 1;
		}
	      else
		dest_backup = NULL;
	    }
	}
    }
  else if (errno != ENOENT)
    {
      error (0, errno, "%s", dest);
      return 1;
    }

  if (verbose)
    printf ("%s -> %s\n", source, dest);

  if (rename (source, dest) == 0)
    {
      return 0;
    }

i=errno;
  if ( i != EEXIST && i != EXDEV)
    {
      error (0, errno, _("cannot move `%s' to `%s'"), source, dest);
      goto un_backup;
    }

  /* rename failed on cross-filesystem link.  Copy the file instead. */

  if (copy_reg (source, dest))
    goto un_backup;

  if (unlink (source))
    {
      error (0, errno, _("cannot remove `%s'"), source);
      return 1;
    }

  return 0;

 un_backup:
  if (dest_backup)
    {
      if (rename (dest_backup, dest))
	error (0, errno, _("cannot un-backup `%s'"), dest);
    }
  return 1;
}

/* Move file SOURCE onto DEST.  Handles the case when DEST is a directory.
   Return 0 if successful, 1 if an error occurred.  */

static int
movefile (const char *source, const char *dest)
{
  strip_trailing_slashes (source);

  if ((dest[strlen (dest) - 1] == '\\' && !is_real_dir (source))
      || isdir (dest))
    {
      /* Target is a directory; build full target filename. */
      char *base;
      char *new_dest;
      int fail;

      base = basename (source);
      new_dest = path_concat (dest, base, NULL);
      if (new_dest == NULL)
	error (1, 0, _("virtual memory exhausted"));
      fail = do_move (source, new_dest);
      free (new_dest);
      return fail;
    }
  else
    return do_move (source, dest);
}

static void
usage (int status)
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      printf (_("\
Usage: %s [OPTION]... SOURCE DEST\n\
  or:  %s [OPTION]... SOURCE... DIRECTORY\n\
"),
	      program_name, program_name);
      printf (_("\
Rename SOURCE to DEST, or move SOURCE(s) to DIRECTORY.\n\
\n\
  -b, --backup                 make backup before removal\n\
  -f, --force                  remove existing destinations, never prompt\n\
  -i, --interactive            prompt before overwrite\n\
  -S, --suffix=SUFFIX          override the usual backup suffix\n\
  -u, --update                 move only older or brand new files\n\
  -v, --verbose                explain what is being done\n\
  -V, --version-control=WORD   override the usual version control\n\
      --help                   display this help and exit\n\
      --version                output version information and exit\n\
\n\
"));
      printf (_("\
The backup suffix is ~, unless set with SIMPLE_BACKUP_SUFFIX.  The\n\
version control may be set with VERSION_CONTROL, values are:\n\
\n\
  t, numbered     make numbered backups\n\
  nil, existing   numbered if numbered backups exist, simple otherwise\n\
  never, simple   always make simple backups\n\
"));
      puts (_("\nReport bugs to fileutils-bugs@gnu.ai.mit.edu"));
    }
  exit (status);
}

int
main (int argc, char **argv)
{
  int c;
  int errors;
  int make_backups = 0;
  char *version;

  program_name = argv[0];
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  version = getenv ("SIMPLE_BACKUP_SUFFIX");
  if (version)
    simple_backup_suffix = version;
  version = getenv ("VERSION_CONTROL");

  myeuid = geteuid ();
  interactive = override_mode = verbose = update = 0;
  errors = 0;

  while ((c = getopt_long (argc, argv, "bfiuvS:V:", long_options, (int *) 0))
	 != EOF)
    {
      switch (c)
	{
	case 0:
	  break;
	case 'b':
	  make_backups = 1;
	  break;
	case 'f':
	  interactive = 0;
	  override_mode = 1;
	  break;
	case 'i':
	  interactive = 1;
	  override_mode = 0;
	  break;
	case 'u':
	  update = 1;
	  break;
	case 'v':
	  verbose = 1;
	  break;
	case 'S':
	  simple_backup_suffix = optarg;
	  break;
	case 'V':
	  version = optarg;
	  break;
	default:
	  usage (1);
	}
    }

  if (show_version)
    {
      printf ("mv (%s) %s\n", GNU_PACKAGE, VERSION);
      exit (0);
    }

  if (show_help)
    usage (0);

  if (argc < optind + 2)
    {
      error (0, 0, "%s", (argc == optind
			  ? _("missing file arguments")
			  : _("missing file argument")));
      usage (1);
    }

  if (make_backups)
    backup_type = get_version (version);

  stdin_tty = isatty (STDIN_FILENO);

  if (argc > optind + 2 && !isdir (argv[argc - 1]))
    error (1, 0,
	   _("when moving multiple files, last argument must be a directory"));

  /* Move each arg but the last onto the last. */
  for (; optind < argc - 1; ++optind)
    errors |= movefile (argv[optind], argv[argc - 1]);

  exit (errors);
}


/* yesno.c -- read a yes/no response from stdin
   Copyright (C) 1990 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <ctype.h>
#ifdef HAVE_STDLIB_H
# include <stdlib.h>
#endif
#include <stdio.h>

/* Read one line from standard input
   and return nonzero if that line begins with y or Y,
   otherwise return 0. */

int rpmatch ();

int
yesno ()
{
  /* We make some assumptions here:
     a) leading white space in the response are not vital
     b) the first 128 characters of the answer are enough (the rest can
	be ignored)
     I cannot think for a situation where this is not ok.  --drepper@gnu  */
  char buf[128];
  int len = 0;
  int c;

  while ((c = getchar ()) != EOF && c != '\n')
    if ((len > 0 && len < 127) || (len == 0 && !isspace (c)))
      buf[len++] = c;
  buf[len] = '\0';
	
  if (*buf=='\0') return 1; // KMS
  return rpmatch (buf) == 1;
}
